File: packages/sdk/android/src/main/java/com/primerioreactnative/datamodels/PrimerCheckoutAdditionalInfoRN.kt
package com.primerioreactnative.datamodels

import kotlinx.serialization.Serializable

@Serializable
sealed class PrimerCheckoutAdditionalInfoRN {
  abstract val additionalInfoName: String
}

@Serializable
data class MultibancoCheckoutAdditionalInfoRN(
  val expiresAt: String,
  val reference: String,
  val entity: String,
) : PrimerCheckoutAdditionalInfoRN() {
  override val additionalInfoName: String = "MultibancoCheckoutAdditionalInfo"
}

sealed class PrimerCheckoutQRCodeInfoRN : PrimerCheckoutAdditionalInfoRN()

@Serializable
data class PromptPayCheckoutAdditionalInfoRN(
  val expiresAt: String,
  val qrCodeUrl: String?,
  val qrCodeBase64: String?,
) : PrimerCheckoutQRCodeInfoRN() {
  override val additionalInfoName: String = "PromptPayCheckoutAdditionalInfo"
}

sealed class PrimerCheckoutVoucherInfoRN : PrimerCheckoutAdditionalInfoRN()

@Serializable
data class XenditCheckoutVoucherAdditionalInfoRN(
  val expiresAt: String,
  val couponCode: String,
  val retailerName: String?,
) : PrimerCheckoutVoucherInfoRN() {
  override val additionalInfoName: String = "XenditCheckoutVoucherAdditionalInfo"
}

@Serializable
class AchAdditionalInfoDisplayMandateRN : PrimerCheckoutVoucherInfoRN() {
  override val additionalInfoName: String = "DisplayStripeAchMandateAdditionalInfo"
}

File: ios/Sources/DataModels/PrimerCheckoutAdditionalInfoRN.swift
//
//  PrimerCheckoutAdditionalInfoRN.swift
//  primer-io-react-native
//
//  Created by Flaviu Dunca on 17.06.2024.
//

import Foundation
import PrimerSDK

protocol PrimerCheckoutAdditionalInfoRN: Encodable {
    var additionalInfoName: String { get }
}

class MultibancoCheckoutAdditionalInfoRN: PrimerCheckoutAdditionalInfoRN {
    let expiresAt: String?
    let reference: String?
    let entity: String?
    let additionalInfoName: String = "MultibancoCheckoutAdditionalInfo"

    init(expiresAt: String?, reference: String?, entity: String?) {
        self.expiresAt = expiresAt
        self.reference = reference
        self.entity = entity
    }
}

protocol PrimerCheckoutQRCodeInfoRN: PrimerCheckoutAdditionalInfoRN {}

class PromptPayCheckoutAdditionalInfoRN: PrimerCheckoutQRCodeInfoRN {
    let expiresAt: String
    let qrCodeUrl: String?
    let qrCodeBase64: String?
    var additionalInfoName: String = "PromptPayCheckoutAdditionalInfo"

    init(expiresAt: String, qrCodeUrl: String?, qrCodeBase64: String?) {
        self.expiresAt = expiresAt
        self.qrCodeUrl = qrCodeUrl
        self.qrCodeBase64 = qrCodeBase64
    }
}

protocol PrimerCheckoutVoucherInfoRN: PrimerCheckoutAdditionalInfoRN {}

class XenditCheckoutVoucherAdditionalInfoRN: PrimerCheckoutVoucherInfoRN {
    let expiresAt: String
    let couponCode: String
    let retailerName: String?
    var additionalInfoName: String = "XenditCheckoutVoucherAdditionalInfo"

    init(expiresAt: String, couponCode: String, retailerName: String?) {
        self.expiresAt = expiresAt
        self.couponCode = couponCode
        self.retailerName = retailerName
    }
}

class AchAdditionalInfoDisplayMandateRN: PrimerCheckoutAdditionalInfoRN {
    var additionalInfoName: String = "DisplayStripeAchMandateAdditionalInfo"
}
